Remove user data from couch user model

esoergel opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

## Technical Summary
This is cleanup from the migration to SQL that happened earlier in the year.  The couch field is no longer being updated and isn't in use anywhere, so this PR removes it.  Turns out removing that field breaks a couple things that accessed the raw user document.  Fixes for those are included.

Since the model change needs to go live before the migration is run, this PR includes a management command to be run post-deploy.  https://github.com/dimagi/commcare-hq/pull/34798 will include a django migration that calls that command for 3rd-party environments.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

There were apparently a handful of places in tests that took issue with this field going away, so there is a chance there are other unknown, untested areas that reference this.  

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->


### Migrations

This has a model change and a management command, but the automatic migration will come in a second PR

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

esoergel committed:
Remove old migration

esoergel committed:
Remove old, unused user_data field

esoergel committed:
This is no longer a class attribute

It is overridden anyways in dehydrate

esoergel committed:
Check user data changes against SQL original

This was previously checking that user data changed before calling
`add_changes`, and then checking again inside `add_changes`, but this
time against the couch field, which shouldn't be in-use anymore

esoergel committed:
Log user data changes and not just from to_json

esoergel committed:
Remove unused user data field

The migration will come later

esoergel committed:
Just use ES results directly

The query only returns docs that _don't_ have that value set, then
fetches those from couch and attempts to get that value.  However, this
information is no longer stored in couch anyways.

esoergel commented on corehq/apps/geospatial/views.py:
```diff
@@ -367,32 +366,28 @@ def get_paginated_cases_without_gps(self, domain, page, limit):
 def _get_paginated_users_without_gps(domain, page, limit, query):
     location_prop_name = get_geo_user_property(domain)
-    query = (
+    res = (
```
@zandre-eng What do you think of this change?

zandre-eng replied:
Apologies for the late response on this as I was away on PTO. The changes look good, and it's nice that we're able to handle pagination now directly in the ES query.

esoergel committed:
Fix related doc lookups for user data

These have been broken since the migration to SQL, apparently

esoergel committed:
Remove db requirement for champ UCR tests

esoergel committed:
Make body of migration run only once

Jtang-1 reviewed (commented):
reviewed, didn't have any additional comments but not enough confidence of the context to approve

esoergel commented on corehq/apps/user_importer/helpers.py:
```diff
@@ -174,20 +174,16 @@ def update_user_data(self, data, uncategorized_data, profile_name, profiles_by_n
             self.logger.add_info(UserChangeMessage.profile_info(user_data.profile_id, profile_name))
 
+        if old_user_data != user_data.raw:
+            self.logger.add_changes({'user_data': user_data.raw}, skip_confirmation=True)
```
Can you elaborate?  This code logs all changes made to a user for auditing purposes.

MartinRiese replied:
Maybe it is just the name of the field being `user_data` and in actuality it does not contain anything sensitive. But I wonder if there was a chance something that we don't want logged ends up in there. I think it would not be clear to every developer changing what goes in `user_data` that it will also end up in the logs.

esoergel replied:
true - though `logger` here actually isn't a normal python logger, it's an instance of `UserChangeLogger`, which saves details of changes to postgres to power an auditing report.

MartinRiese replied:
Fair

esoergel commented on corehq/apps/api/resources/v0_1.py:
```diff
@@ -61,7 +61,7 @@ class Meta(CustomResourceMeta):
 class CommCareUserResource(UserResource):
     groups = fields.ListField(attribute='get_group_ids')
-    user_data = fields.DictField(attribute='user_data')
+    user_data = fields.DictField()
```
Yep, good question.  That's handled here:
https://github.com/dimagi/commcare-hq/blob/d3d0b803705c56aae41c68d5fc53aba2129e29d1/corehq/apps/api/user_updates.py#L112-L122

esoergel commented on corehq/apps/users/util.py:
```diff
@@ -398,19 +398,20 @@ def log_user_change(by_domain, for_domain, couch_user, changed_by_user, changed_
 
     if action in [UserModelAction.CREATE, UserModelAction.DELETE]:
         changed_details = couch_user.to_json()
+        changed_details['user_data'] = couch_user.get_user_data(for_domain).raw if for_domain else {}
```
that's right :grimacing:

esoergel commented on corehq/apps/geospatial/views.py:
```diff
@@ -367,32 +366,28 @@ def get_paginated_cases_without_gps(self, domain, page, limit):
         .sort('created_on', desc=True)
+        .start((page - 1) * limit)
+        .size(limit)
+        .run()
```
yep.  And then fetching the users again from couch.  It does the same thing in the function below this one too - I just didn't want to get carried away with cleanup.

esoergel commented on corehq/apps/userreports/expressions/specs.py:
```diff
@@ -600,19 +600,21 @@ def __call__(self, item, evaluation_context=None):
+        if domain != doc.get('domain'):
             return None
+        if related_doc_type == 'CommCareUser':
+            doc['user_data'] = CommCareUser.wrap(doc).get_user_data(domain).to_dict()
```
> I guess user data just hasn't been part of UCRs since the migration?

That's right - this has been broken since the migration.

You can't do a related doc lookup on a web user, since they don't have a top level `domain` attribute.  For the same reason, they also don't have a `user_data` attribute (since it varies from one domain to the next), though I guess we could inject one in this context.

esoergel replied:
For what it's worth, this means of directly accessing the `user_data` from the raw JSON has been broken to some degree for years - user data is only supposed to be accessed via helper methods, as fields controlled by the profile aren't stored directly in the user document.  That means that the profile wasn't accounted for in the API or in UCR.  I just broke it further :laughing:

esoergel commented on corehq/dbaccessors/couchapps/all_docs.py:
```diff
@@ -37,6 +37,18 @@ def get_doc_count_by_type(db, doc_type):
+    return (row['id'] for row in paginate_view(
+        db,
+        'all_docs/by_doc_type',
+        chunk_size=10000,
```
There's no particular diligence behind this number - just figured with only IDs I'd go higher than for docs.

esoergel commented on corehq/apps/users/management/commands/rm_couch_user_data.py:
```diff
@@ -0,0 +1,31 @@
+                 + get_doc_count_by_type(db, 'CommCareUser'))
+        all_ids = chain(iter_all_doc_ids(db, 'WebUser'),
+                        iter_all_doc_ids(db, 'CommCareUser'))
+        iter_update(db, _update_user, with_progress_bar(all_ids, count), verbose=True)
```
[iter_update](https://github.com/dimagi/commcare-hq/blob/959fd0dd5fb61f5539db795d14e61178dc8b6d3a/corehq/util/couch.py#L264-L264) is akin to `map` - it fetches documents by ID in chunks, applies `_update_user` to them, and saves the return value (also in chunks) if anything changed.  So here `_update_user` defines the actual change to each document, which is `user_doc.pop('user_data', ...)` (`.pop` differs from `.get` in that it mutates the object).

esoergel commented on corehq/apps/users/management/commands/rm_couch_user_data.py:
```diff
@@ -0,0 +1,31 @@
+
+
+class Command(BaseCommand):
+    help = "Populate SQL user data from couch"
```
oh whoops you're right - I forgot to update the help text.  Let me do that now

esoergel committed:
update help text

esoergel commented:
Just ran the management command on staging, where it succeeded without issues:
```
Elapsed time: 0:28:04
couldn't find 0 docs
ignored 86 docs
deleted 0 docs
updated 76144 docs
```

esoergel commented:
Ran the management command on this PR on prod:
```
Started at 2024-07-16 19:26:20
Processing [..................................................] 1395446/1395444 100% 1:02:32.965644 elapsed
Finished at 2024-07-16 20:28:53
Elapsed time: 1:02:33
```

india:
```
Started at 2024-07-16 19:27:33
Processing [..................................................] 159858/159858 100% 0:18:35.724473 elapsed
Finished at 2024-07-16 19:46:09
Elapsed time: 0:18:36
```

and swiss:
```
Started at 2024-07-16 19:28:49
Processing [..................................................] 935/935 100% 0:00:02.457244 elapsed
Finished at 2024-07-16 19:28:51
Elapsed time: 0:00:02
```
